File: go/server/middleware.go
package server

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"

	mpp "github.com/solana-foundation/pay-kit/go"
	"github.com/solana-foundation/pay-kit/go/errorcodes"
)

type contextKey string

const receiptContextKey contextKey = "mpp-receipt"

func markAuthorizationBoundResponse(header http.Header) {
	header.Set("Cache-Control", "no-store")

	for _, value := range header.Values("Vary") {
		for _, field := range strings.Split(value, ",") {
			if strings.EqualFold(strings.TrimSpace(field), "Authorization") || strings.TrimSpace(field) == "*" {
				return
			}
		}
	}
	header.Add("Vary", "Authorization")
}

// ReceiptFromContext extracts the payment receipt from the request context.
func ReceiptFromContext(ctx context.Context) (mpp.Receipt, bool) {
	r, ok := ctx.Value(receiptContextKey).(mpp.Receipt)
	return r, ok
}

// ChargeFunc returns the charge amount and options for a given request.
type ChargeFunc func(r *http.Request) (amount string, opts ChargeOptions, err error)

// PaymentMiddleware wraps an http.Handler to enforce MPP payments.
// Requests without a valid credential get a 402 challenge.
// Requests with a valid credential get the receipt injected into context.
func PaymentMiddleware(m *Mpp, chargeFn ChargeFunc) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Serve the service worker JS when the query param is present.
			if IsServiceWorkerRequest(r) {
				w.Header().Set("Content-Type", "application/javascript")
				w.WriteHeader(http.StatusOK)
				_, _ = w.Write([]byte(ServiceWorkerJS()))
				return
			}

			// Resolve the route's expected charge first so verification can be
			// route-aware: the credential's claimed amount is compared to this
			// route's expected amount, not just to itself.
			amount, opts, err := chargeFn(r)
			if err != nil {
				http.Error(w, "charge function error", http.StatusInternalServerError)
				return
			}

			challenge, err := m.ChargeWithOptions(r.Context(), amount, opts)
			if err != nil {
				http.Error(w, "failed to create challenge", http.StatusInternalServerError)
				return
			}

			// Check for a payment credential in the Authorization header.
			// verificationErr captures whichever verification step failed
			// so the 402 body can surface a canonical L6 code. A nil
			// verificationErr on the re-challenge path means the caller
			// never sent an Authorization header (or sent an empty one).
			var verificationErr error
			authHeader := r.Header.Get(mpp.AuthorizationHeader)
			if paymentToken, ok := mpp.ExtractPaymentScheme(authHeader); ok && paymentToken != "" {
				credential, err := mpp.ParseAuthorization(authHeader)
				if err != nil {
					verificationErr = mpp.WrapError(mpp.ErrCodeInvalidPayload, "parse authorization", err)
				} else {
					var expected mpp.ChargeRequest
					if decodeErr := challenge.Request.Decode(&expected); decodeErr != nil {
						verificationErr = mpp.WrapError(mpp.ErrCodeInvalidPayload, "decode challenge request", decodeErr)
					} else {
						receipt, verifyErr := m.VerifyCredentialWithExpected(r.Context(), credential, expected)
						if verifyErr == nil {
							receiptHeader, fmtErr := mpp.FormatReceipt(receipt)
							if fmtErr == nil {
								w.Header().Set(mpp.PaymentReceiptHeader, receiptHeader)
							}
							markAuthorizationBoundResponse(w.Header())
							ctx := context.WithValue(r.Context(), receiptContextKey, receipt)
							next.ServeHTTP(w, r.WithContext(ctx))
							return
						}
						verificationErr = verifyErr
					}
				}
			}

			wwwAuth, err := mpp.FormatWWWAuthenticate(challenge)
			if err != nil {
				http.Error(w, "failed to format challenge", http.StatusInternalServerError)
				return
			}

			w.Header().Set(mpp.WWWAuthenticateHeader, wwwAuth)
			markAuthorizationBoundResponse(w.Header())

			if m.HTMLEnabled() && AcceptsHTML(r) {
				html, err := m.ChallengeToHTML(challenge)
				if err == nil {
					w.Header().Set("Content-Type", "text/html; charset=utf-8")
					w.WriteHeader(http.StatusPaymentRequired)
					_, _ = w.Write([]byte(html))
					return
				}
				// Fall through to JSON on HTML error.
			}

			code := errorcodes.PaymentInvalid
			message := "Payment required"
			if verificationErr != nil {
				code = errorcodes.CanonicalFromError(verificationErr)
				message = verificationErr.Error()
			}
			body, err := json.Marshal(errorcodes.NewPaymentRequiredBody(code, message))
			if err != nil {
				http.Error(w, "failed to marshal challenge body", http.StatusInternalServerError)
				return
			}
			w.Header().Set("Content-Type", "application/problem+json")
			w.WriteHeader(http.StatusPaymentRequired)
			_, _ = w.Write(body)
		})
	}
}

File: python/src/solana_pay_kit/protocols/mpp/server/middleware.py
"""Payment decorator for ASGI/Starlette-style handlers."""

from __future__ import annotations

import functools
from collections.abc import Callable
from typing import Any

from solana_pay_kit._paycore.errors import PaymentError, payment_required_response
from solana_pay_kit.protocols.mpp.core.headers import format_www_authenticate, parse_authorization
from solana_pay_kit.protocols.mpp.server.charge import Mpp


def pay(mpp_handler: Mpp, amount: str, **options: Any) -> Callable:
    """Decorator for ASGI/Starlette-style handlers.

    Wraps a handler to automatically handle 402 Payment Required flows.
    The decorated handler receives (request, credential, receipt) when
    payment is verified.

    Example:
        @app.get("/paid")
        @pay(mpp, amount="0.50")
        async def handler(request, credential, receipt):
            return {"data": "paid content"}
    """
    from solana_pay_kit.protocols.mpp.intents.charge import ChargeRequest
    from solana_pay_kit.protocols.mpp.server.charge import ChargeOptions

    charge_options = ChargeOptions(
        description=options.get("description", ""),
        external_id=options.get("external_id", ""),
        expires=options.get("expires", ""),
        fee_payer=options.get("fee_payer", False),
        splits=options.get("splits", []),
    )

    def decorator(handler: Callable) -> Callable:
        @functools.wraps(handler)
        async def wrapper(request: Any, *args: Any, **kwargs: Any) -> Any:
            # Build the route's expected charge first so verification can be
            # route-aware: the credential's claimed amount is compared to this
            # route's expected amount, not just to itself.
            challenge = mpp_handler.charge_with_options(amount, charge_options)

            # Try to get Authorization header
            auth_header = None
            if hasattr(request, "headers"):
                auth_header = request.headers.get("authorization")

            verification_error: PaymentError | None = None
            if auth_header:
                try:
                    credential = parse_authorization(auth_header)
                    expected = ChargeRequest.from_dict(challenge.decode_request())
                    receipt = await mpp_handler.verify_credential_with_expected(credential, expected)
                    return await handler(request, credential, receipt, *args, **kwargs)
                except PaymentError as err:
                    verification_error = err
                except Exception as err:  # noqa: BLE001 (catch-all for framework parse errors)
                    verification_error = PaymentError(str(err), code="payment_invalid")

            # Issue (or re-issue) a 402 with a canonical L6 code in the body.
            www_auth = format_www_authenticate(challenge)
            if verification_error is None:
                response = payment_required_response(
                    "Payment required",
                    code="payment_invalid",
                    challenge_header=www_auth,
                )
            else:
                response = payment_required_response(
                    str(verification_error) or "Payment required",
                    code=verification_error.code or "payment_invalid",
                    challenge_header=www_auth,
                )
            response["__mpp_challenge"] = True
            response["challenge"] = challenge
            return response

        return wrapper

    return decorator
